feat: Add relational filter conditions in data browser

vardhan0604 opened the pull request:
### New Pull Request Checklist
<!--
    Please check the following boxes [x] before submitting your issue.
    Click the "Preview" tab for better readability.
    Thanks for contributing to Parse Server!
-->

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-dashboard/issues/2496).

### Issue Description
<!-- Add a brief description of the issue this PR solves. -->

Added filter conditions that relate to other classes. Since Parse Dashboard knows the entire schema, it also knows which other classes have a pointer to the currently displayed class; it can use that information to allow for filtering by fields of these related classes.

![image](https://github.com/parse-community/parse-dashboard/assets/91624754/5072771e-37c4-43ed-a94c-cff012609386)

Closes: #2496

vardhan0604 committed:
feat:Add relational filter conditions (#2496)

vardhan0604 committed:
Merge branch 'alpha' of https://github.com/vardhan0604/parse-dashboard into alpha

vardhan0604 committed:
undo changes in  Parse-Dashboard/parse-dashboard-config.json

vardhan0604 commented:
@mtrezza, can you please check this

vardhan0604 committed:
fixes lint issue

vardhan0604 committed:
undo changes in parse-dashboard-config.json

vardhan0604 commented:
Hey @mtrezza, I have fixed the lint issue, can you please review the PR

mtrezza committed:
Merge branch 'alpha' into alpha

mtrezza commented:
The dashboard data browser seems to be continuously loading the table; not sure if that is a bug related to this feature, still investigating. You can see this by launching the preview environment from the uffizzi-cloud comment and just opening a collection like "_User".

Edit: I can confirm that the issue only occurs in this PR; could you look into that?

mtrezza commented on src/dashboard/Data/Browser/Browser.react.js:
```diff
@@ -101,9 +101,13 @@ class Browser extends DashboardView {
       draggedRowSelection: false,
+
+      classes: {},
+      AllclassesSchema: {}
```
```suggestion
      allClassesSchema: {}
```

mtrezza commented on src/dashboard/Data/Browser/Browser.react.js:
```diff
@@ -101,9 +101,13 @@ class Browser extends DashboardView {
     };
 
     this.addLocation = this.addLocation.bind(this);
+    this.AllclassesSchema = this.getAllclassesSchema.bind(this);
```
```suggestion
    this.allClassesSchema = this.getAllClassesSchema.bind(this);
```

mtrezza commented on src/dashboard/Data/Browser/BrowserToolbar.react.js:
```diff
@@ -225,14 +227,39 @@ const BrowserToolbar = ({
     });
   }
 
+  Allclasses.forEach(className => {
```
```suggestion
  allClasses.forEach(className => {
```

mtrezza reviewed (changes requested):
I've marked some var names with incorrect caps. There are more, could you please correct them?

vardhan0604 commented:
> I've marked some var names with incorrect caps. There are more, could you please correct them?

> data browser seems to be continuously loading the table

Sure, I will correct them and I will look into the second issue as well

mtrezza commented:
The preview environment does not update on every commit. If you want to update the environment you can close and re-open this PR, then after a few minutes uffizzi will post a new comment with the update env link.

vardhan0604 commented:
> The preview environment does not update on every commit. If you want to update the environment you can close and re-open this PR, then after a few minutes uffizzi will post a new comment with the update env link.

ohh okay, I will keep that in my mind

vardhan0604 committed:
corrected the variable name using right naming conventions

vardhan0604 committed:
fixing continous loading of the table

vardhan0604 commented:
> data browser seems to be continuously loading the table

@mtrezza I have fixed this issue and also fixed the var names.
can you please review the code?

mtrezza committed:
delete file

Signed-off-by: Manuel <5673677+mtrezza@users.noreply.github.com>

mtrezza commented on src/components/BrowserFilter/BrowserFilter.react.js:
```diff
@@ -46,13 +46,18 @@ export default class BrowserFilter extends React.Component {
       );
-      const field = Object.keys(available)[0];
-      filters = new List([new Map({ field: field, constraint: available[field][0] })]);
+      const filterClass = Object.keys(available)[0];
+      const filterField = Object.keys(available[filterClass])[0];
+      const filterConstraint = available[filterClass][filterField][0];
+      filters = new List([
+        new Map({ class: filterClass, field: filterField, constraint: filterConstraint }),
+      ]);
```
This code block seems to occur almost identical in `addRow`. Could you refactor these occurrences to deduplicate? Maybe keep the `availableFilters` method and move the code block there?

vardhan0604 replied:
> Maybe keep the availableFilters method

It's still there `findRelatedClasses` uses the availableFilters method

maybe i can make a function like this 
```
export function getFilterDetails(available) {
  const filterClass = Object.keys(available)[0];
  const filterField = Object.keys(available[filterClass])[0];
  const filterConstraint = available[filterClass][filterField][0];
  return { filterClass, filterField, filterConstraint };
}
```

and call it here like

```
addRow() {
    const available = Filters.findRelatedClasses(
      this.props.className,
      this.props.allClassesSchema,
      this.state.blacklistedFilters,
      this.state.filters
    );
    const { filterClass, filterField, filterConstraint } = Filters.getFilterDetails(available);
    this.setState(({ filters }) => ({
      filters: filters.push(
        new Map({ class: filterClass, field: filterField, constraint: filterConstraint })
      ),
      editMode: true,
    }));
  }
  ```
  if you think this works , i can push the changes

mtrezza replied:
I haven't looked at the logic; I just notice that there is duplicate code which maybe could be optimized. If your suggestion works and it removes the duplication, please go ahead.

vardhan0604 replied:
actually the duplicate code is nothing just getting the class , field and constraint value from the available. 

I made a function which takes the available and returns a object which contain of class , field and this makes the code more neat in my take and removes the duplication

mtrezza commented on src/components/Filter/Filter.react.js:
```diff
@@ -12,28 +12,58 @@ import React from 'react';
+  const current = filters.get(index);
+  const field = current.get('field');
+  const constraint = current.get('constraint');
+
+  const newClassFields = Object.keys(schema[newClassName]);
+  const isFieldValid = newClassFields.includes(field);
+
+  const newField = isFieldValid ? field : newClassFields[0];
+  const allowedConstraints = Filters.FieldConstraints[schema[newClassName][newField].type];
+  const isConstraintValid = allowedConstraints.includes(constraint);
+
+  const newConstraint = isConstraintValid ? constraint : allowedConstraints[0];
+
+  const defaultCompare = Filters.DefaultComparisons[schema[newClassName][newField].type];
+
+  const newFilter = new Map({
```
```suggestion
  const newClassFields = Object.keys(schema[newClassName]);
  const isFieldValid = newClassFields.includes(field);
  const newField = isFieldValid ? field : newClassFields[0];
  const allowedConstraints = Filters.FieldConstraints[schema[newClassName][newField].type];
  const isConstraintValid = allowedConstraints.includes(constraint);
  const newConstraint = isConstraintValid ? constraint : allowedConstraints[0];
  const defaultCompare = Filters.DefaultComparisons[schema[newClassName][newField].type];
  const newFilter = new Map({
```

mtrezza committed:
remove empty lines

Signed-off-by: Manuel <5673677+mtrezza@users.noreply.github.com>

vardhan0604 committed:
removing duplicacy

vardhan0604 committed:
Merge branch 'alpha' of https://github.com/vardhan0604/parse-dashboard into alpha

mtrezza reviewed (commented):
Code looks good; will do some test scenarios to see if it breaks anything else.

vardhan0604 commented:
> Code looks good; will do some test scenarios to see if it breaks anything else.

sure lemme know if i need to make any changes

mtrezza commented:
I tested the [example scenario](https://github.com/parse-community/parse-dashboard/issues/2496#issue-1811823229) described in the feature request and it worked well. I can already tell that this feature will be very helpful to many dashboard users, good job.  Here are just some minor details I have found:

- When adding a new filter condition, the default class should be the currently visible class in the data browser. In my test it defaulted to the custom `Transaction` class despite adding the filter being in the `_User` class.
- For better visual, could you add a margin between the "Clear" and "Add" buttons? This also prevents accidentally clicking "Clear" instead of "Add". FYI, I've added a 150px margin-right to the "Clear" button in the image.
   <img width="690" alt="image" src="https://github.com/parse-community/parse-dashboard/assets/5673677/544add2f-e76a-497b-9db2-50fc1ec9e4f4">
- Could you please add an entry to the README.md under "Features" > "Data Browser" > "Filter" in which you describe in a few words that the data browser allows to add relational filter conditions based on other classes that have a pointer to the current class. Basically, one should understand that a class only shows up if it has a pointer to the class that is currently displayed.

Note:
- I tested saving the filter and applying the saved filter, both worked fine.

vardhan0604 committed:
added the default class as the current class in the filter

vardhan0604 committed:
updated filter dialog UI

vardhan0604 commented:
- >When adding a new filter condition, the default class should be the currently visible class in the data browser. In my test it defaulted to the custom Transaction class despite adding the filter being in the _User class.

- >For better visual, could you add a margin between the "Clear" and "Add" buttons? This also prevents accidentally clicking "Clear" instead of "Add". FYI, I've added a 150px margin-right to the "Clear" button in the image.

- >Could you please add an entry to the README.md under "Features" > "Data Browser" > "Filter" in which you describe in a few words that the data browser allows to add relational filter conditions based on other classes that have a pointer to the current class. Basically, one should understand that a class only shows up if it has a pointer to the class that is currently displayed.


Hey @mtrezza , I have done the first two changes asked by you but for the 3rd I'm not sure where and how to do that?
![image](https://github.com/parse-community/parse-dashboard/assets/91624754/05537f5b-75ea-4b2e-9a5b-01e11ae01edb)
is this how we are supposed to do or how?
sorry i havent contributed in docs before

vardhan0604 committed:
updated the readme.md

mtrezza commented:
Just one last thing, given that our filter dialog is now more complex with an additional column, do you think you could add the headings "Class, Field, Condition, Value", like in the [screenshot](https://github.com/parse-community/parse-dashboard/issues/2496#issue-1811823229)? I believe this would make the UI easier to use. If that's too much work, no worries, we can merge without.

And I found another tiny issue: classes with a pointer to the current class only appear in the filter dialog after reloading the dashboard in the browser window. It seems that the schema is not updated locally after adding a pointer field to a class. I'm not sure whether this is a general dashboard issue that the schema is not updated, or whether it is related to this feature.

mtrezza committed:
Update README.md

vardhan0604 committed:
fixed new class not showing locally

vardhan0604 committed:
Merge branch 'alpha' of https://github.com/vardhan0604/parse-dashboard into alpha

vardhan0604 committed:
updated the filter dialog UI

vardhan0604 commented:
> Just one last thing, given that our filter dialog is now more complex with an additional column, do you think you could add the headings "Class, Field, Condition, Value", like in the [screenshot](https://github.com/parse-community/parse-dashboard/issues/2496#issue-1811823229)? I believe this would make the UI easier to use. If that's too much work, no worries, we can merge without.
> 
> And I found another tiny issue: classes with a pointer to the current class only appear in the filter dialog after reloading the dashboard in the browser window. It seems that the schema is not updated locally after adding a pointer field to a class. I'm not sure whether this is a general dashboard issue that the schema is not updated, or whether it is related to this feature.

Hey @mtrezza , I have fixed the second issue and added the headings in the UI.  you can check it and test it out.
lemme know if something breaks

mtrezza reviewed (commented):
It seems that the class is still only displaying after a website refresh. Could you check please? The headings look great!

vardhan0604 committed:
fix new class not loading without reload

vardhan0604 commented:
> It seems that the class is still only displaying after a website refresh. Could you check please? The headings look great!

Hey @mtrezza , you can check I have fixed the issue

mtrezza reviewed (dismissed):
Looks good! Thanks for this PR!

mtrezza committed:
lint fix

parseplatformorg commented:
🎉 This change has been released in version [6.0.0-alpha.9](https://github.com/parse-community/parse-dashboard/releases/tag/6.0.0-alpha.9)

parseplatformorg commented:
🎉 This change has been released in version [6.0.0-beta.2](https://github.com/parse-community/parse-dashboard/releases/tag/6.0.0-beta.2)

parseplatformorg commented:
🎉 This change has been released in version [6.0.0](https://github.com/parse-community/parse-dashboard/releases/tag/6.0.0)
